File: epi_judge_cpp/buy_and_sell_stock.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;
double BuyAndSellStockOnce(const vector<double>& prices) {
  double max_gain=0, min_so_far = std::numeric_limits<double>::max();
  for(double price: prices) {
    double gain = price - min_so_far;
    max_gain = std::max(max_gain, gain);
    min_so_far = std::min(min_so_far, price);
  }
  return max_gain;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"prices"};
  return GenericTestMain(args, "buy_and_sell_stock.cc",
                         "buy_and_sell_stock.tsv", &BuyAndSellStockOnce,
                         DefaultComparator{}, param_names);
}

File: epi_judge_python/buy_and_sell_stock.py
from typing import List
import sys

from test_framework import generic_test


def buy_and_sell_stock_once(prices: List[float]) -> float:
    max_gain = 0
    min_so_far = sys.maxsize
    for price in prices:
        gain = price - min_so_far
        max_gain = max(max_gain, gain)
        min_so_far = min(min_so_far, price)
    return max_gain


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('buy_and_sell_stock.py',
                                       'buy_and_sell_stock.tsv',
                                       buy_and_sell_stock_once))
